### ISSUE TITLE
traefik memory usage keeps growing

### ISSUE THREAD

**Issue by ravilr** [containous/traefik#4864]
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [* ] No

### What did you do?

we are notificing RSS memory of traefik process running within Kubernetes pod container keeps growing, eventually reaching the configured memory limits and getting OOMKilled.  

heap dump pprof profile shows most of the memory accumulated at ' crypto/tls.(*block).reserve /usr/local/go/src/crypto/tls/conn.go'
```
(pprof) top20
Showing nodes accounting for 982.10MB, 96.97% of 1012.75MB total
Dropped 227 nodes (cum <= 5.06MB)
Showing top 20 nodes out of 97
      flat  flat%   sum%        cum   cum%
  740.49MB 73.12% 73.12%   740.49MB 73.12%  crypto/tls.(*block).reserve /usr/local/go/src/crypto/tls/conn.go
   73.56MB  7.26% 80.38%    73.56MB  7.26%  crypto/tls.Server /usr/local/go/src/crypto/tls/tls.go
   47.01MB  4.64% 85.02%    47.01MB  4.64%  crypto/aes.(*aesCipherGCM).NewGCM /usr/local/go/src/crypto/aes/aes_gcm.go
   30.51MB  3.01% 88.04%    30.51MB  3.01%  crypto/aes.newCipher /usr/local/go/src/crypto/aes/cipher_asm.go
   15.50MB  1.53% 89.57%    15.50MB  1.53%  crypto/tls.(*halfConn).newBlock /usr/local/go/src/crypto/tls/conn.go
   13.50MB  1.33% 90.90%    13.50MB  1.33%  bytes.makeSlice /usr/local/go/src/bytes/buffer.go
      13MB  1.28% 92.18%       13MB  1.28%  syscall.anyToSockaddr /usr/local/go/src/syscall/syscall_linux.go
      12MB  1.19% 93.37%       12MB  1.19%  net.newFD /usr/local/go/src/net/fd_unix.go
      11MB  1.09% 94.45%   126.60MB 12.50%  crypto/tls.(*Conn).readHandshake /usr/local/go/src/crypto/tls/conn.go
       8MB  0.79% 95.24%        8MB  0.79%  crypto/tls.ekmFromMasterSecret /usr/local/go/src/crypto/tls/prf.go
       7MB  0.69% 95.94%        7MB  0.69%  net.sockaddrToTCP /usr/local/go/src/net/tcpsock_posix.go
    5.52MB  0.55% 96.48%     5.52MB  0.55%  bufio.NewWriterSize /usr/local/go/src/bufio/bufio.go
    3.50MB  0.35% 96.83%    81.02MB  8.00%  crypto/tls.aeadAESGCM /usr/local/go/src/crypto/tls/cipher_suites.go
    1.50MB  0.15% 96.97%   123.60MB 12.20%  crypto/tls.(*serverHandshakeState).readClientHello /usr/local/go/src/crypto/tls/handshake_server.go
         0     0% 96.97%   214.84MB 21.21%  bufio.(*Reader).Peek /usr/local/go/src/bufio/bufio.go
         0     0% 96.97%    28.21MB  2.79%  bufio.(*Reader).Read /usr/local/go/src/bufio/bufio.go
         0     0% 96.97%    57.20MB  5.65%  bufio.(*Reader).ReadLine /usr/local/go/src/bufio/bufio.go
         0     0% 96.97%    57.20MB  5.65%  bufio.(*Reader).ReadSlice /usr/local/go/src/bufio/bufio.go
         0     0% 96.97%   272.04MB 26.86%  bufio.(*Reader).fill /usr/local/go/src/bufio/bufio.go
         0     0% 96.97%   125.74MB 12.42%  bufio.(*Writer).Flush /usr/local/go/src/bufio/bufio.go
```
![profile001](https://user-images.githubusercontent.com/3928411/57489993-ab733780-726c-11e9-98c4-fd2323ed0fbf.png)

### What did you expect to see?

No unlimited memory growth/leak

### What did you see instead?
memory leak

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
traefik v1.7.11
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

running on Kubernetes 
```toml
    # Enable debug mode
    #
    # Optional
    # Default: false
    #
    debug = true

    # Periodically check if a new version has been released.
    checkNewVersion = false

    # Log level
    #
    # Optional
    # Default: "ERROR"
    #
    logLevel = "WARN"

    # Entrypoints to be used by frontends that do not specify any entrypoint.
    # Each frontend can specify its own entrypoints.
    #
    # Optional
    # Default: ["http"]
    #
    defaultEntryPoints = ["http","https"]

    # If set to true invalid SSL certificates are accepted for backends.
    # This disables detection of man-in-the-middle attacks so should only be used on secure backend networks.
    #
    # Optional
    # Default: false
    InsecureSkipVerify = true

    # Entrypoints definition
    #
    # Optional
    # Default:
    [entryPoints]
      [entryPoints.http]
      address = ":20080"

      [entryPoints.https]
      address = ":20443"
        [entryPoints.https.tls]

      [entryPoints.traefik-admin]
      address = "127.0.0.1:28080"

    ### Ping configuration
    [ping]
    entrypoint = "http"

    ### Metrics configuration
    [metrics]
      # To enable Traefik to export internal metrics to Prometheus
      [metrics.prometheus]
        # Buckets for latency metrics
        buckets=[0.1,0.3,1.2,5.0]
        # Name of the related entry point
        entrypoint = "traefik-admin"
      # Enable more detailed statistics.
      [metrics.statistics]
        # Number of recent errors logged.
        recentErrors = 10

    # Traefik logs
    [traefikLog]
      filePath = "/var/log/traefik.log"
      format = "json"

    # Enable access logs
    [accessLog]
      # Sets the file path for the access log. If not specified, stdout will be used.
      # Intermediate directories are created if necessary.
      filePath = "/var/log/traefik-access.log"
      format = "json"
      bufferingSize = 100
        [accessLog.filters]
          statusCodes = ["500-504"]
        [accessLog.fields]
          defaultMode = "keep"
          [accessLog.fields.names]
            "ClientUsername" = "drop"
            "StartUTC" = "drop"
          [accessLog.fields.headers]
            defaultMode = "drop"
            [accessLog.fields.headers.names]
              "User-Agent" = "keep"
              "X-Forwarded-For" = "keep"

    [lifeCycle]
      # Duration to keep accepting requests prior to initiating the graceful
      # termination period (as defined by the `graceTimeOut` option). This
      # option is meant to give downstream load-balancers sufficient time to
      # take Traefik out of rotation.
      requestAcceptGraceTimeout = "25s"

      # Duration to give active requests a chance to finish before Traefik stops.
      graceTimeOut = "15s"

    [forwardingTimeouts]
      # dialTimeout is the amount of time to wait until a connection to a backend server can be established.
      dialTimeout = "5s"

      # responseHeaderTimeout is the amount of time to wait for a server's response headers after fully writing the request (including its body, if any).
      responseHeaderTimeout = "180s"

    ##############################################
    # api configuration backend
    ##############################################
    [api]
      # Name of the related entry point
      entryPoint = "traefik-admin"

      # Enabled Dashboard
      dashboard = true

      # Enable more detailed statistics.
      [api.statistics]
        # Number of recent errors logged.
        recentErrors = 10

    ################################################################
    # Kubernetes Ingress configuration backend
    ################################################################

    # Enable Kubernetes Ingress configuration backend.
    [kubernetes]

    # Kubernetes server endpoint.
    #
    # Optional for in-cluster configuration, required otherwise.
    # Default: empty
    #
    # endpoint = "http://localhost:8080"

    # Bearer token used for the Kubernetes client configuration.
    #
    # Optional
    # Default: empty
    #
    # token = "my token"

    # Path to the certificate authority file.
    # Used for the Kubernetes client configuration.
    #
    # Optional
    # Default: empty
    #
    # certAuthFilePath = "/my/ca.crt"

    # Array of namespaces to watch.
    #
    # Optional
    # Default: all namespaces (empty array).
    #
    # namespaces = ["default", "production"]

    # Ingress label selector to identify Ingress objects that should be processed.
    #
    # Optional
    # Default: empty (process all Ingresses)
    #
    # labelselector = "A and not B"

    # Disable PassHost Headers.
    #
    # Optional
    # Default: false
    #
    # disablePassHostHeaders = true
```

Not sure if this is related, there is some websockets based requests usage and we notice errors like this in the error log:

```
{"level":"error","msg":"vulcand/oxy/forward/websocket: Error dialing \"app1....com\": websocket: bad handshake with resp: 404 404 Not Found","time":"2019-05-09T21:39:51Z"}
```

**Comment by ravilr** [comment]
there was a similar issue reported here https://github.com/golang/go/issues/28654  and it was closed referring to an issue with gorilla request context.  
It looks like traefik does shallow copying of http.request object  and that is known to cause issues like this( when gorilla packages are in dependency) :
https://github.com/vulcand/oxy/blob/master/forward/fwd.go#L473

**Comment by ravilr** [comment]
/debug/pprof/heap output here: https://github.com/ravilr/pprof/raw/master/heap11.pprof

**Comment by ravilr** [comment]
one more data point: we weren't seeing this memory leak even with websocket requests and similar loads. It seems like the leak only manifests when there are failed websocket requests (404 from upstream service) .

**Comment by ravilr** [comment]
@juliens this indeed turns out to be a traefik bug.  Any failed websocket hijacked connections are not cleaned up on close from traefik's hijackConnectionTracker. this is in turn keeping all those tls conn references forever, manifesting as memory leak. 

i tested with this patch below, and don't observe this memory leak:

```
diff --git a/vendor/github.com/vulcand/oxy/forward/fwd.go b/vendor/github.com/vulcand/oxy/forward/fwd.go
index f70ba324..31c44e41 100644
--- a/vendor/github.com/vulcand/oxy/forward/fwd.go
+++ b/vendor/github.com/vulcand/oxy/forward/fwd.go
@@ -363,7 +363,13 @@ func (f *httpForwarder) serveWebSocket(w http.ResponseWriter, req *http.Request,
                                ctx.errHandler.ServeHTTP(w, req, errHijack)
                                return
                        }
-                       defer conn.Close()
+
+                       defer func() {
+                               conn.Close()
+                               if f.websocketConnectionClosedHook != nil {
+                                       f.websocketConnectionClosedHook(req, conn)
+                               }
+                       }()
 
                        errWrite := resp.Write(conn)
                        if errWrite != nil {
```

**Comment by juliens** [comment]
Thanks for your interest in Traefik.

Good catch for this leak, don't hesitate to create a pull request on oxy repository.

**Comment by ravilr** [comment]
@juliens  Can i make another pr against traefik-1.7 branch to dep update oxy here.

**Comment by ldez** [comment]
@ravilr  yes

**Comment by dtomcej** [comment]
Closed by #4900
